Race condition on text resources processed in multiple project instances

We have periodically seen this failure in our own build:

```
src\Tasks\system.design\system.design.txt(0,0): error MSB3103: Invalid Resx file. The process cannot access the file 'D:\a\1\s\src\Tasks\system.design\system.design.txt' because it is being used by another process.
```

Today I looked into it. I suspect this flow of things:

https://github.com/dotnet/msbuild/blob/797fd829a374a880f63fd4eea4ea6990404c48f4/src/Tasks/GenerateResource.cs#L3561-L3565

https://github.com/dotnet/msbuild/blob/797fd829a374a880f63fd4eea4ea6990404c48f4/src/Tasks/GenerateResource.cs#L3876-L3877

`File.Open` does not specify `FileShare.Read`, which defaults to "no sharing". So if this is run on the same file in parallel (for instance, in our Tasks project which is multitargeted), there's a possibility that the second attempt to read the file fails because the first one is still holding a lock.

--- a/src/Tasks/GenerateResource.cs
+++ b/src/Tasks/GenerateResource.cs
@@ -3874,7 +3874,7 @@ namespace Microsoft.Build.Tasks
             private int _col;
 
             internal LineNumberStreamReader(String fileName, Encoding encoding, bool detectEncoding)
-                : base(File.Open(fileName, FileMode.Open, FileAccess.Read), encoding, detectEncoding)
+                : base(File.Open(fileName, FileMode.Open, FileAccess.Read, FileShare.Read), encoding, detectEncoding)
             {
                 _lineNumber = 1;
                 _col = 0;

